app/modules/asset_groups/models.py

```diff
@@ -585,7 +585,7 @@ def __repr__(self):
 
     @property
     def bulk_upload(self):
-        return 'uploadType' in self.config and self.config['uploadType'] == 'bulk'
+        return isinstance(self.config, dict) and self.config.get('uploadType') == 'bulk'
```

Comment:
This change is fine but there is an underlying problem that this is highlighting.
self.config should never be None and this is something I want to investigate

Reply:
It happened in a test like this:

```
git checkout origin/main
git cherry-pick origin/tests-avoid-recloning  # Pick only the "Avoid recloning of asset group fixtures in tests" commit
pytest -s -x --random-order-seed=688303
```

```
tests/modules/asset_groups/resources/test_commit_asset_group.py::test_commit_individual_asset_group[None] FAILED

==================================================================== FAILURES =====================================================================
____________________________________________________ test_commit_individual_asset_group[None] _____________________________________________________

flask_app_client = <AutoAuthFlaskClient <Flask 'app'>>
researcher_1 = <User(guid=f0775a83-44fd-47a7-99bf-6d7e2846227e, email="researcher1@localhost", name="First Middle Last", state=Active, Alpha, roles=Contributor, Researcher)>
regular_user = <User(guid=bbc974d1-5322-4dbc-814e-2722d80eba01, email="test@localhost", name="First Middle Last", state=Active, Alpha, roles=Contributor)>
staff_user = <User(guid=dce39137-de39-4bd6-b831-6904be59a1a8, email="staff@localhost", name="First Middle Last", state=Active, Alpha, roles=Staff, Contributor)>
test_root = PosixPath('/code/tests/asset_groups/test-000'), db = <SQLAlchemy engine=postgresql://houston:***@db/houston_test>
empty_individual = <Individual(guid=899eddf4-97a6-4174-bf67-cb2bd6318be1, )>

    def test_commit_individual_asset_group(
        flask_app_client,
        researcher_1,
        regular_user,
        staff_user,
        test_root,
        db,
        empty_individual,
    ):  
        # pylint: disable=invalid-name                           
        from app.modules.sightings.models import Sighting, SightingStage
                                                                 
        transaction_id, test_filename = asset_group_utils.create_bulk_tus_transaction(
            test_root                                            
        )                                                        
        asset_group_uuid = None                                  
        sighting_uuid = None                                     
        try:                                                     
            data = asset_group_utils.get_bulk_creation_data(transaction_id, test_filename)
            with db.session.begin():                             
                db.session.add(empty_individual)                 
            data.set_encounter_field(0, 0, 'individualUuid', str(empty_individual.guid))
                                                                 
>           resp = asset_group_utils.create_asset_group(         
                flask_app_client, researcher_1, data.get()       
            )                                                    
                                                                 
tests/modules/asset_groups/resources/test_commit_asset_group.py:209:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/modules/asset_groups/resources/utils.py:83: in create_asset_group
    response = flask_app_client.post(                            
/usr/local/lib/python3.9/site-packages/werkzeug/test.py:1039: in post
    return self.open(*args, **kw)                                
tests/utils.py:87: in open                                       
    response = super(AutoAuthFlaskClient, self).open(*args, **kwargs)
/usr/local/lib/python3.9/site-packages/flask/testing.py:222: in open
    return Client.open(                                          
/usr/local/lib/python3.9/site-packages/werkzeug/test.py:993: in open
    response = self.run_wsgi_app(environ.copy(), buffered=buffered)
/usr/local/lib/python3.9/site-packages/werkzeug/test.py:884: in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
/usr/local/lib/python3.9/site-packages/werkzeug/test.py:1119: in run_wsgi_app
    app_rv = app(environ, start_response)                        
/usr/local/lib/python3.9/site-packages/flask/app.py:2464: in __call__
    return self.wsgi_app(environ, start_response)                
/usr/local/lib/python3.9/site-packages/flask/app.py:2450: in wsgi_app
    response = self.handle_exception(e)                          
/usr/local/lib/python3.9/site-packages/flask_cors/extension.py:165: in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
/usr/local/lib/python3.9/site-packages/flask_restx/api.py:671: in error_router
    return original_handler(f)                                   
/usr/local/lib/python3.9/site-packages/flask/app.py:1867: in handle_exception
    reraise(exc_type, exc_value, tb)                             
/usr/local/lib/python3.9/site-packages/flask/_compat.py:39: in reraise
    raise value                                                  
/usr/local/lib/python3.9/site-packages/flask_restx/api.py:669: in error_router
    return self.handle_error(e)                                  
/usr/local/lib/python3.9/site-packages/flask/app.py:2447: in wsgi_app
    response = self.full_dispatch_request()                      
/usr/local/lib/python3.9/site-packages/flask/app.py:1952: in full_dispatch_request
    rv = self.handle_user_exception(e)                           
/usr/local/lib/python3.9/site-packages/flask_cors/extension.py:165: in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
/usr/local/lib/python3.9/site-packages/flask_restx/api.py:671: in error_router
    return original_handler(f)                                   
/usr/local/lib/python3.9/site-packages/flask/app.py:1821: in handle_user_exception
    reraise(exc_type, exc_value, tb)                             
/usr/local/lib/python3.9/site-packages/flask/_compat.py:39: in reraise
    raise value                                                  
/usr/local/lib/python3.9/site-packages/flask_restx/api.py:669: in error_router
    return self.handle_error(e)                                  
/usr/local/lib/python3.9/site-packages/flask/app.py:1950: in full_dispatch_request
    rv = self.dispatch_request()                                 
/usr/local/lib/python3.9/site-packages/flask/app.py:1936: in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)   
/usr/local/lib/python3.9/site-packages/flask_restx/api.py:403: in wrapper
    resp = resource(*args, **kwargs)                             
/usr/local/lib/python3.9/site-packages/flask/views.py:89: in view
    return self.dispatch_request(*args, **kwargs)                
/usr/local/lib/python3.9/site-packages/flask_restx/resource.py:49: in dispatch_request
    resp = meth(*args, **kwargs)                                 
flask_restx_patched/namespace.py:397: in wrapper                 
    return func(*args, **kwargs)                                 
/usr/local/lib/python3.9/site-packages/webargs/core.py:459: in wrapper
    return func(*new_args, **kwargs)                             
flask_restx_patched/namespace.py:145: in dump_wrapper            
    response = func(*args, **kwargs)                             
app/modules/asset_groups/resources.py:79: in post                
    metadata.process_request()                                   
app/modules/asset_groups/metadata.py:366: in process_request
    self._validate_contents()                                    
app/modules/asset_groups/metadata.py:408: in _validate_contents  
    if group.bulk_upload and not group.is_processed():           
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
                                                                 
self = <AssetGroup(guid=00000000-0000-0000-0000-000000000003, )> 
                                                                 
    @property                                                    
    def bulk_upload(self):                                       
>       return 'uploadType' in self.config and self.config['uploadType'] == 'bulk'
E       TypeError: argument of type 'NoneType' is not iterable   
                                                                 
app/modules/asset_groups/models.py:588: TypeError                
```

Reply:
I think what is happening here is when you're doing a bulk upload, the code checks that the current user doesn't have any other bulk uploads:

https://github.com/WildMeOrg/houston/blob/2678a80a7955379aa4ae2e7356c8908d6366961e/app/modules/asset_groups/metadata.py#L406-L412

The AssetGroup model itself is ok with `config` being null:

https://github.com/WildMeOrg/houston/blob/2678a80a7955379aa4ae2e7356c8908d6366961e/app/modules/asset_groups/models.py#L561

So it all depends on the view code which of course the fixtures aren't using...

May be it's better to add a default like AssetGroupSighting.jobs:

https://github.com/WildMeOrg/houston/blob/2678a80a7955379aa4ae2e7356c8908d6366961e/app/modules/asset_groups/models.py#L125

Reply:
Thanks for suggestions, I had a bit of an investigation today and taken them on board along with a few other fixes.